src/block/Furnace.php

```diff
@@ -90,4 +93,77 @@ public function onScheduledUpdate() : void{
 			$world->scheduleDelayedBlockUpdate($this->position, 1); //TODO: check this
 		}
 	}
+
+	public function doHopperPull(Hopper $hopperBlock) : bool{
+		$currentTile = $this->position->getWorld()->getTile($this->position);
+		if(!$currentTile instanceof TileFurnace){
+			return false;
+		}
+
+		$tileHopper = $this->position->getWorld()->getTile($hopperBlock->position);
+		if(!$tileHopper instanceof TileHopper){
+			return false;
+		}
+
+		$sourceInventory = $tileHopper->getInventory();
+		$targetInventory = $currentTile->getInventory();
+
+		$hopperFacing = $hopperBlock->getFacing();
+
+		for($i = 0; $i < $targetInventory->getSize(); $i++){
+			$itemStack = $sourceInventory->getItem($i);
+
+			if($itemStack->isNull()){
+				continue;
+			}
+
+			$singleItem = $itemStack->pop(1);
+
+			if($hopperFacing === Facing::DOWN && $targetInventory->canAddSmelting($singleItem)){
+				$this->transferItem($sourceInventory, $targetInventory, $singleItem, $targetInventory::SLOT_INPUT);
```

Comment:
It's debatable, but I believe it would be more readable to use `FurnaceInventory::SLOT_INPUT` here than `$targetInventory::SLOT_INPUT`.